### PR TITLE
* Tidy up `DefineFonts()`

### PR DESCRIPTION
**Pull request by Wagnerp**
* Tidy up `DefineFonts()`

image

### CODE REVIEW

File: Source/Krypton Components/Krypton.Toolkit/Palette Builtin/Base/PaletteBase.cs [hunk 1]
```diff
@@ -1231,7 +1231,15 @@ public virtual float BaseFontSize
         {
             get => _baseFont ?? new("Segoe UI", BaseFontSize, FontStyle.Regular);
```
Comment (Smurf-IV) [inline Source/Krypton Components/Krypton.Toolkit/Palette Builtin/Base/PaletteBase.cs:1232]: This means that the return is never null.
Therefore remove the nullable return type

File: Source/Krypton Components/Krypton.Toolkit/Palette Builtin/Base/PaletteBase.cs [hunk 2]
```diff
@@ -1231,7 +1231,15 @@ public virtual float BaseFontSize
...
 
-            set => _baseFont = value;
+            set
+            {
+                _baseFont = value;
+
+                if (value == null)
+                {
+                    _baseFont = new("Segoe UI", BaseFontSize, FontStyle.Regular);
```
Comment (Smurf-IV) [inline Source/Krypton Components/Krypton.Toolkit/Palette Builtin/Base/PaletteBase.cs:1240]: This means that the code can accept a null but not return it (See above) so mark with `[AllowNull]`

**Review by Smurf-IV — CHANGES_REQUESTED** [review]

File: Source/Krypton Components/Krypton.Toolkit/Palette Builtin/Microsoft 365/Bases/PaletteMicrosoft365Base.cs [hunk 3]
```diff
@@ -4900,41 +4900,90 @@ public virtual string BaseFontName
         protected override void DefineFonts()
         {
             // Release existing resources
-            _header1ShortFont?.Dispose();
-
-            _header2ShortFont?.Dispose();
+            if (_header1ShortFont != null)
```
Comment (Smurf-IV) [inline Source/Krypton Components/Krypton.Toolkit/Palette Builtin/Microsoft 365/Bases/PaletteMicrosoft365Base.cs:4903]: Why has this been turned into the long format. That is what the `?.` does for the code.

**Review by Smurf-IV — CHANGES_REQUESTED** [review]

**Comment by Smurf-IV** [comment]
I thought the idea of this code change was to move the Font dispose stuff into a base class, but it appears that this has not been done.

**Comment by Wagnerp** [comment]
> I thought the idea of this code change was to move the Font dispose stuff into a base class, but it appears that this has not been done.

@Smurf-IV That will be done today. Looks like **all** font stuff will need to be moved.

File: Source/Krypton Components/Krypton.Toolkit/Palette Builtin/Base/PaletteBase.cs (hunk 1 above)
Comment (Wagnerp) [reply Source/Krypton Components/Krypton.Toolkit/Palette Builtin/Base/PaletteBase.cs:1232]: Done

File: Source/Krypton Components/Krypton.Toolkit/Palette Builtin/Base/PaletteBase.cs (hunk 2 above)
Comment (Wagnerp) [reply Source/Krypton Components/Krypton.Toolkit/Palette Builtin/Base/PaletteBase.cs:1240]: Done

**Comment by Wagnerp** [comment]
@Smurf-IV I've now split the `DefineFonts` function up into two in the base class. Though I'm not sure if `virtual` is the right type to use?

```cs
/// <summary>
        /// Update the fonts to reflect system or user defined changes.
        /// </summary>
        protected virtual void DefineFonts()
        {
            // Release resources
            DisposeFonts();

            if (_baseFont != null)
            {
                _header1ShortFont = new Font(_baseFont.Name, _baseFont.Size + 4.5f, FontStyle.Bold);
                _header2ShortFont = new Font(_baseFont.Name, _baseFont.Size, FontStyle.Regular);
                _headerFormFont = new Font(_baseFont.Name, SystemFonts.CaptionFont.SizeInPoints, FontStyle.Regular);
                _header1LongFont = new Font(_baseFont.Name, _baseFont.Size + 1.5f, FontStyle.Regular);
                _header2LongFont = new Font(_baseFont.Name, _baseFont.Size, FontStyle.Regular);
                _buttonFont = new Font(_baseFont.Name, _baseFont.Size, FontStyle.Regular);
                _buttonFontNavigatorStack = new Font(_buttonFont, FontStyle.Bold);
                _buttonFontNavigatorMini = new Font(_baseFont.Name, _baseFont.Size + 3.5f, FontStyle.Bold);
                _tabFontNormal = new Font(_baseFont.Name, _baseFont.Size, FontStyle.Regular);
                _tabFontSelected = new Font(_tabFontNormal, FontStyle.Bold);
                _ribbonTabFont = new Font(_baseFont.Name, _baseFont.Size, FontStyle.Regular);
                _gridFont = new Font(_baseFont.Name, _baseFont.Size, FontStyle.Regular);
                _superToolFont = new Font(_baseFont.Name, _baseFont.Size, FontStyle.Bold);
                _calendarFont = new Font(_baseFont.Name, _baseFont.Size, FontStyle.Regular);
                _calendarBoldFont = new Font(_baseFont.Name, _baseFont.Size, FontStyle.Bold);
                _boldFont = new Font(_baseFont.Name, _baseFont.Size, FontStyle.Bold);
                _italicFont = new Font(_baseFont.Name, _baseFont.Size, FontStyle.Italic);
            }
        }

        protected virtual void DisposeFonts()
        {
            _header1ShortFont?.Dispose();
            _header2ShortFont?.Dispose();
            _header1LongFont?.Dispose();
            _header2LongFont?.Dispose();
            _headerFormFont?.Dispose();
            _buttonFont?.Dispose();
            _buttonFontNavigatorStack?.Dispose();
            _buttonFontNavigatorMini?.Dispose();
            _tabFontNormal?.Dispose();
            _tabFontSelected?.Dispose();
            _ribbonTabFont?.Dispose();
            _gridFont?.Dispose();
            _calendarFont?.Dispose();
            _calendarBoldFont?.Dispose();
            _superToolFont?.Dispose();
            _boldFont?.Dispose();
            _italicFont?.Dispose();
        }
```

**Comment by Smurf-IV** [comment]
>I've now split the DefineFonts function up into two in the base class. 

The `_baseFont` is never going to be null, as it is default initialised to something via the accessor (So use that into a local variable.).

> Though I'm not sure if virtual is the right type to use?

It allows someone to add new fonts in the future (For a theme) and still call the base when needed

**Comment by Wagnerp** [comment]
Getting there... think I'm now on the right track!!!

File: Source/Krypton Components/Krypton.Toolkit/Palette Builtin/Microsoft 365/Bases/PaletteMicrosoft365Base.cs (hunk 3 above)
Comment (Wagnerp) [reply Source/Krypton Components/Krypton.Toolkit/Palette Builtin/Microsoft 365/Bases/PaletteMicrosoft365Base.cs]: Moved to base

File: Source/Krypton Components/Krypton.Toolkit/Palette Builtin/Base/PaletteBase.cs [hunk 4]
```diff
@@ -26,16 +26,30 @@ public abstract class PaletteBase : Component
         private Padding? _inputControlPadding;
         private PaletteDragFeedback _dragFeedback;
         private string? _themeName;
-        private Font? _baseFont;
 
-        #endregion
+        #region Font Stuff
 
-        #region Public
+        private Font? _baseFont;
```
Comment (Smurf-IV) [inline Source/Krypton Components/Krypton.Toolkit/Palette Builtin/Base/PaletteBase.cs:32]: `baseFont` does not need to be nullable, as it is never access outside this class, and is set on first use of Get accesser

**Review by Smurf-IV — CHANGES_REQUESTED** [review]

File: Source/Krypton Components/Krypton.Toolkit/Palette Builtin/Base/PaletteBase.cs (hunk 4 above)
Comment (Wagnerp) [reply Source/Krypton Components/Krypton.Toolkit/Palette Builtin/Base/PaletteBase.cs]: Removed flag

**Pull request merged**